app/src/main/java/com/amaze/filemanager/ui/activities/MainActivityViewModel.kt

```diff
@@ -72,4 +82,96 @@ class MainActivityViewModel(val applicationContext: Application) :
     fun getFromMediaFilesCache(mediaType: Int): List<LayoutElementParcelable>? {
         return mediaCacheHash[mediaType]
     }
+
+    /**
+     * Perform basic search: searches on the current directory
+     */
+    fun basicSearch(mainActivity: MainActivity, query: String):
+        MutableLiveData<ArrayList<HybridFileParcelable>> {
+        val hybridFileParcelables = ArrayList<HybridFileParcelable>()
+
+        val mutableLiveData:
+            MutableLiveData<ArrayList<HybridFileParcelable>> =
+            MutableLiveData(hybridFileParcelables)
+
+        val showHiddenFiles = PreferenceManager
+            .getDefaultSharedPreferences(mainActivity)
+            .getBoolean(PREFERENCE_SHOW_HIDDENFILES, false)
+
+        viewModelScope.launch(Dispatchers.IO) {
+            listFiles(
+                mainActivity.currentMainFragment!!.currentPath!!,
+                mainActivity.isRootExplorer,
+                showHiddenFiles,
+                { _: OpenMode? -> null }
+            ) { hybridFileParcelable: HybridFileParcelable ->
+                if (hybridFileParcelable.getName(mainActivity)
+                    .lowercase(Locale.getDefault())
+                    .contains(query.lowercase(Locale.getDefault())) &&
+                    (showHiddenFiles || !hybridFileParcelable.isHidden)
+                ) {
+                    hybridFileParcelables.add(hybridFileParcelable)
+
+                    mutableLiveData.postValue(hybridFileParcelables)
+                }
+            }
+        }
+
+        return mutableLiveData
+    }
+
+    /**
+     * Perform indexed search: on MediaStore items from the current directory & it's children
+     */
+    fun indexedSearch(
+        mainActivity: MainActivity,
+        query: String
+    ): MutableLiveData<ArrayList<HybridFileParcelable>> {
+        val list = ArrayList<HybridFileParcelable>()
+
+        val mutableLiveData: MutableLiveData<ArrayList<HybridFileParcelable>> = MutableLiveData(
+            list
+        )
+
+        val showHiddenFiles =
+            PreferenceManager.getDefaultSharedPreferences(mainActivity)
+                .getBoolean(PREFERENCE_SHOW_HIDDENFILES, false)
+
+        viewModelScope.launch(Dispatchers.IO) {
+            val projection = arrayOf(MediaStore.Files.FileColumns.DATA)
+
+            val cursor = mainActivity
+                .contentResolver
+                .query(MediaStore.Files.getContentUri("external"), projection, null, null, null)
+                ?: return@launch
```

Comment:
@VishalNehra please disable "Expression with labels increase complexity and affect maintainability." on Codacy. In the case of this PR, I can't remove that label, or else the code wouldn't compile!